text/0000-Bundle-Load-API.md

```diff
@@ -51,18 +52,26 @@ The output of the API will be the function pointer type supplied by the API user
 * 
 * @return A function pointer symbols of type T supplied by the user or nullptr if the library is not loaded
 * 
-* @throws std::runtime_error if this symbol fetched is nullptr
+* @throws std::runtime_error if the bundle is not started and active
+* @throws std::runtime_error if the symbol fetched is nullptr
+*
+* @pre  Bundle is already started and active
 *
 * @post The symbol(s) associated with the bundle gets fetched if the library is loaded
+* @post If the symbol does not exist, the API returns nullptr 
 */
 
 template<typename T>
 T GetSymbol(const std::string& symname);
 
 ```
-This API will interally call SharedLibrary class's API to retrieve handle of bundle's shared library. 
-It will then supply the name and type of the symbol and fetch it from bundle class's utility functions.
+This API will first check if the bundle (from which the symbols are to be fetched) is in START + ACTIVE state.
+If it is, then it will internally call SharedLibrary class's API GetHandle to retrieve the handle of bundle's shared library.
+If not, it will throw an exception with message as "invalid bundle".
```

Comment:
A DS bundle's shared library is not loaded into the process when its `Bundle::Start()` method is called. The bundle's shared library should only be loaded the first time a client attempts to get a service provided by the bundle. 

How will `Bundle::GetSymbol` know when to correctly attempt to load the shared library into the process?

Reply:
@jeffdiclemente I think the way it will work is by checking if the `SharedLibrary` object returns true for `IsLoaded()` or not. If no, then we can call the `Load()` API on SharedLibrary with options supplied from `LIBRARY_LOAD_OPTIONS` (as mentioned by saschazelzer)

Either this or DS can call `SharedLibrary::Load()` from its implementation (as mentioned by saschazelzer), after which we can conclude if the shared library is loaded or not.

Reply:
@gitabhishekdp I went back and walked the code path that DS takes to load the shared library and I think the simplest thing to do which satisfies the requirements is to
- call `SharedLibrary::Load` from `GetComponentCreatorDeletors`. 
- call the `Bundle::GetSymbol` function from `GetComponentCreatorDeletors` to get the required function symbols.
Since DS listens for bundle started events, DS attempts to load the shared library from `GetComponentCreatorDeletors` only if it received a bundle started event (and all of the bundle's service dependencies are satisfied). 

`Bundle::GetSymbol` should never attempt to load a shared library. I'd suggest keeping it simple and having DS make two API calls to `SharedLibrary::Load` and `Bundle::GetSymbol`.

Reply:
@jeffdiclemente Thanks for the explanation. I will make the relevant changes to the document.